src/compose/update-strategies.ts

```diff
@@ -12,12 +13,30 @@ export interface StrategyContext {
 	dependenciesMetForStart: boolean;
 	dependenciesMetForKill: boolean;
 	needsSpecialKill: boolean;
+	currentServices: Service[];
+	currentServicesLocked: boolean;
+	appsToLock: AppsToLockMap;
+}
+
+function generateLockThenKillStep(
+	current: Service,
+	currentServices: Service[],
+	currentServicesLocked: boolean,
+	appsToLock: AppsToLockMap,
+): Nullable<CompositionStep> {
+	if (!currentServicesLocked) {
+		currentServices.forEach((svc) =>
+			appsToLock[svc.appId].add(svc.serviceName),
+		);
+		return;
```

Comment:
Why return undefined vs a `noop` step here?

Reply:
I don't think we need to add to the number of loops the state funnel has to go through with a noop. `appsToLock` later gets parsed into a single takeLock step, the way for it to remain a single takeLock step and not a bunch of disjointed takeLock steps is for a map to be passed around between `generateStepsForService` for each service, which tracks which services require the lock.